### PR TITLE
feat(tree): approximate range row-count + selectivity

### LINKED ISSUE #505: Approximate range stats: add row-count cardinality + zone-map selectivity for join ordering
> [!WARNING]
> **On-disk format is pre-stable.** There are no external consumers yet, so the V5 disk format is **not frozen** and is expected to change through the format issues (#502, #503, #505, #507, #508). While pre-stable, **old on-disk data is not supported across format changes** — there is no backward-compat reader and no migration code. Regenerate the data instead (the existing per-block checksum / AEAD rejects a stale layout with a clear error rather than misreading it). Format freeze + a stable on-disk version + migration discipline begin with the first external consumer.

---

## Problem

`approximate_range_stats` (#499) as specified is byte-based (RocksDB `GetApproximateSizes` style). Byte size alone is insufficient for join ordering: the planner needs an estimate of the NUMBER OF ROWS in a range and the SELECTIVITY of a predicate, not just bytes.

## Contract

- Add an approximate **row-count** estimate per range (e.g. per-block row-counts maintained in the index), in addition to byte size.
- Add predicate **selectivity** estimation over a range using the per-column zone-maps from #502 (fraction of blocks / rows that can match a given range or equality predicate).

## Acceptance

- The row-count estimate is within a bounded error of the true count on a synthetic dataset.
- The selectivity estimate monotonically reflects predicate tightness (a tighter predicate yields a lower estimate), verified on a controlled dataset.

## Dependency

#502 (per-column zone maps). Extends #499.

## no-std tier

Estimation read-path = core + alloc.

## Estimate

~1 week.

### PR DESCRIPTION
**Pull request by polaz**
## Summary

Adds `AbstractTree::approximate_range_cardinality(range, seqno) -> RangeCardinality { rows, selectivity }`: a row-count and selectivity estimate for a key range, computed **without reading any data block**. For cost-based query planning (join ordering, scan-vs-seek).

## How it works

- **rows**: each overlapping SST's per-data-block zone-map (#502) row counts are summed for every block whose key range overlaps the query (block granularity). A table written without a zone map falls back to the byte-fraction row estimate (the `approximate_range_stats` path). The active + sealed memtables add their in-range, snapshot-visible counts.
- **selectivity**: `rows / total_rows`, in `0.0..=1.0`, monotonic in predicate tightness.
- Snapshot/visibility, per-table seqno translation, and comparator-aware overlap match the read path (and `approximate_range_stats`).

The zone-map reader (`ZoneMap::columns_for`) now has a live caller, so its dead-code suppression on the table field is removed; the cursor API stays gated for the future columnar block-skip scan (#503).

## Tests

6 integration tests: full range = exact total + selectivity 1.0; empty range = 0; sub-range row count within a bounded (block-granularity) error of an actual scan; selectivity monotonic across widening ranges; memtable-only counted; fallback path without a zone map. Plus a `# Examples` doctest. Full suite (2056), `clippy --all-features --all-targets`, no-std (`alloc`), and doctests green.

## Stacking

Branched from `feat/#499-approximate-range-stats` (PR #513) since it extends `approximate_range_stats`. Base retargets to `main` after #513 merges.

Closes #505

## Release Notes

* **New Features**
  * Added key-range cardinality estimation to the public tree API, returning estimated row counts and selectivity for cost-based query planning at a specified snapshot.
  * Uses zone-map metadata when available; otherwise falls back to a byte-fraction approximation while maintaining monotonic selectivity as ranges widen.
  * Supported for SSTs and memtables, including restricted-view behavior.

* **Documentation**
  * Added and exported `RangeCardinality` to expose the estimation results.

* **Tests**
  * Added integration tests covering zone-map precision, fallback behavior, empty/full ranges, monotonicity, and boundary cases.

### CODE REVIEW

File: src/tree/mod.rs [hunk 1]
```diff
@@ -1169,6 +1169,178 @@ impl AbstractTree for Tree {
...
+                    if idx_bytes > 0 {
+                        let est = u64::try_from(
+                            u128::from(table.metadata.item_count) * u128::from(idx_bytes)
+                                / u128::from(data_end),
+                        )
+                        .unwrap_or(u64::MAX)
+                        .max(1);
+                        rows = rows.saturating_add(est);
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/tree/mod.rs]: Fixed — gated on the zone map's presence (`!zone_map.is_empty()`) instead of the shared counted flag. When the zone map is present it is now authoritative: a range that lands in a key-space gap reports zero rows and never falls back to the byte fraction; only a truly absent zone map falls back. Regression test `range_in_a_data_gap_with_zone_map_is_zero` (one key per block so no block straddles the gap).

File: src/tree/mod.rs [hunk 2]
```diff
@@ -1169,6 +1169,173 @@ impl AbstractTree for Tree {
...
+                // minimum key is above the upper bound; the boundary block at `lo`
+                // is counted in full (block granularity). A range that lands in a
+                // key-space gap legitimately yields zero, so this path is
+                // authoritative and never falls back to the byte fraction.
+                let reader = match lo {
+                    Bound::Included(k) | Bound::Excluded(k) => {
+                        table.block_index.forward_reader(k, table_seqno)
+                    }
+                    Bound::Unbounded => Some(table.block_index.iter()),
+                };
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/tree/mod.rs]: @coderabbitai Fixed — both branches now raise the table's effective lower bound to `restrict_lower_bound()` (a per-table `eff_lo` = max(lo, restriction)), so a restricted view's punched-out prefix no longer contributes rows, mirroring `approximate_range_stats`.
